circuits/plonk/src/constraints.rs

```diff
@@ -0,0 +1,173 @@
+/*****************************************************************************************************************
+
+This source file implements Plonk circuit constraint primitive.
+
+*****************************************************************************************************************/
+
+use algebra::{FftField, SquareRootField};
+use ff_fft::{Evaluations, EvaluationDomain, Radix2EvaluationDomain as Domain, GeneralEvaluationDomain, DensePolynomial};
+pub use super::gate::CircuitGate;
+use rand_core::OsRng;
+
+#[derive(Clone)]
+pub struct ConstraintSystem<F: FftField>
+{
+    pub public: usize,                      // number of public inputs
+    pub domain: Domain<F>,                  // evaluation domain
+    pub gates:  Vec<CircuitGate<F>>,        // circuit gates
+
+    // index polynomials over the monomial base
+    pub sigmam: [DensePolynomial<F>; 3],    // permutation polynomial array
+    pub ql:     DensePolynomial<F>,         // left input wire polynomial
+    pub qr:     DensePolynomial<F>,         // right input wire polynomial
+    pub qo:     DensePolynomial<F>,         // output wire polynomial
+    pub qm:     DensePolynomial<F>,         // multiplication polynomial
+    pub qc:     DensePolynomial<F>,         // constant wire polynomial
+
+    pub sigmal: [Vec<F>; 3],                // permutation polynomial array over Lagrange base
+    pub sid:    Vec<F>,                     // SID polynomial over Lagrange base
+    pub r:      F,                          // coordinate shift for right wires
+    pub o:      F,                          // coordinate shift for output wires
+}
+
+impl<F: FftField + SquareRootField> ConstraintSystem<F> 
+{
+    pub fn evals(v : Vec<F>, d : Domain<F>) -> Evaluations<F, GeneralEvaluationDomain<F>>
+    {
+        Evaluations::<F>::from_vec_and_domain(v, GeneralEvaluationDomain::Radix2(d))
+    }
+
+    pub fn evals_from_coeffs(&self, v : Vec<F>) -> Evaluations<F, GeneralEvaluationDomain<F>>
+    {
+        Evaluations::<F>::from_vec_and_domain(v, GeneralEvaluationDomain::Radix2(self.domain))
+    }
+
+    pub fn create
+    (
+        mut gates: Vec<CircuitGate<F>>,
+        public: usize,
+    ) -> Option<Self>
+    {
+        let domain = Domain::<F>::new(Domain::<F>::compute_size_of_domain(gates.len())?)?;
+        let mut sid = domain.elements().map(|elm| {elm}).collect::<Vec<_>>();
+        let r =
+        {
+            let mut r = domain.sample_element_outside_domain(&mut OsRng);
+            while r.legendre().is_qnr() == false {r = domain.sample_element_outside_domain(&mut OsRng)}
```

Comment:
It might be better to pass down an implementer of an Rng trait rather than using OsRng directly.